ipld/retriever.go

```diff
@@ -0,0 +1,178 @@
+package ipld
+
+import (
+	"context"
+	"math/rand"
+	"time"
+
+	"github.com/ipfs/go-cid"
+	format "github.com/ipfs/go-ipld-format"
+	logging "github.com/ipfs/go-log/v2"
+	"github.com/ipfs/go-merkledag"
+	"github.com/tendermint/tendermint/pkg/da"
+	"github.com/tendermint/tendermint/pkg/wrapper"
+
+	"github.com/celestiaorg/celestia-node/ipld/plugin"
+	"github.com/celestiaorg/nmt"
+	"github.com/celestiaorg/rsmt2d"
+)
+
+var log = logging.Logger("ipld")
+
+// RetrieveQuadrantTimeout limits the time for retrieval of a quadrant
+// so that Retriever can retry another quadrant.
+var RetrieveQuadrantTimeout = time.Minute * 5
+
+// Retriever retrieves rsmt2d.ExtendedDataSquares from the IPLD network.
+// Instead of requesting data 'share by share' it requests data by quadrants
+// minimizing bandwidth usage in the happy cases.
+//
+//  ---- ----
+// | 0  | 1  |
+//  ---- ----
+// | 2  | 3  |
+//  ---- ----
+// Retriever randomly picks one of the data square quadrants and tries to request them one by one until it is able
+// the whole square.
+type Retriever struct {
+	dag   format.DAGService
+	codec rsmt2d.Codec
+}
+
+// NewRetriever creates a new instance of the Retriever over IPLD Service and rmst2d.Codec
+func NewRetriever(dag format.DAGService, codec rsmt2d.Codec) *Retriever {
+	return &Retriever{dag: dag, codec: codec}
+}
+
+// Retrieve retrieves all the data committed to DataAvailabilityHeader.
+// It aims to request from the network only 1/4 of the data and reconstructs other 3/4 parts using the rsmt2d.Codec.
+// It steadily tries to request other 3/4 data if 1/4 is not found within RetrieveQuadrantTimeout or unrecoverable.
+// TODO(@Wondertan): Randomize requesting from Col Roots with retrying.
+func (r *Retriever) Retrieve(ctx context.Context, dah *da.DataAvailabilityHeader) (*rsmt2d.ExtendedDataSquare, error) {
+	daroots := dah.RowsRoots
```

Comment:
Q: Can you remind me why we only sample from rows?

Reply:
There is no reason. We should be ideally fetch randomly from either cols or rows. There is a todo mentioning this, and there is even an issue for retries from col. I just made it consistent with the original implementation.

Reply:
I updated the PR to include randomization and retrying for cols